### PR TITLE
authorize_membership_application for all MembershipApplicationsController actions

### PR DESCRIPTION
**Pull request by weedySeaDragon**
## PT Story:  authorize current_user for all actions with a MembershipApplication
https://www.pivotaltracker.com/story/show/153526466

Doing this uncovered 2 issues:
1. A Member has no way to submit a 2nd application.  I created a PT feature for this: https://www.pivotaltracker.com/story/show/153724872
2. On the MembershipApplication, there was a guard clause on the `start_review` event that  would not let the application state transition from `` to `` if the user is a member.  That guard clause would also prevent a Member from submitting a 2nd (new) application.  I removed the guard clause.

## Changes proposed in this pull request:
1. MembershipApplicationsController
   -  change from limited :only... checking certain actions to _always_ checking authorization for _all_ actions
   - `check_and_mark_if_ready_for_review` changed to be a private method
2. MembershipApplicationPolicy: 
   - fixed logic for `new?` check (an admin cannot create a new application because we do not have a way for the admin to specify the User it is for)
   - `create?` hardened logic to handle either a Class or instance
   - added check for `index` action
   - added check for `start_review` action
   - improved readability (method names; corrected constant name that wasn't right)
2.  The MembershipApplicationPolicy spec was incomplete.  It did not fully test for all possible actions, and it did not check for all possible combinations of user type and MembershipApplication state (ex:  `under_review,' 'waiting_for_applicant', 'accepted', 'rejected', etc.)   So made sure to check for all the possibilities.  Then I refactored to DRY up the spec.
3. ApplicationPolicy: 
   - added `new?`
   - `admin_or_owner?` : when checking for ownership, the record _must_ be able to respond to `:user` (otherwise it cannot have an "owner")
5. Added a spec for ApplicationPolicy

Ready for review:
@AgileVentures/shf-project-team

### CODE REVIEW

**Review by thesuss — APPROVED** [review]

File: app/policies/shf_application_policy.rb [hunk 1]
```diff
@@ -1,30 +1,35 @@
 class ShfApplicationPolicy < ApplicationPolicy
 
-  EDITABLE_STATES_FOR_APPLICANT = Set[:new, :initial, :ready_for_review, :under_review, :waiting_for_applicant].freeze
+  EDITABLE_STATES_FOR_APPLICATION = Set[:new, :initial, :ready_for_review, :under_review, :waiting_for_applicant].freeze
 
 
   def permitted_attributes
-    allowed_attribs_for_current_user
+    allowed_changable_attribs_for_current_user
```
Comment (patmbolger) [inline app/policies/shf_application_policy.rb]: `changeable`

**Review by patmbolger — APPROVED** [review]
Nicely done!

**Comment by RobertCram** [comment]
The story for the second application already exists (https://www.pivotaltracker.com/story/show/148243533), so we should probably close that?

**Review by RobertCram — APPROVED** [review]

**Comment by weedySeaDragon** [comment]
@RobertCram That story is about the 'larger' feature.  This is just about _authorization_.  
So when it comes time to implement that story, this little piece will (hopefully) already be accomplished.

**Pull request merged**